### ISSUE TITLE
Mark Cases as OT

### ISSUE THREAD

**Issue by araposo-tistatech** [department-of-veterans-affairs/caseflow#13415]
## User or job story
User story: As a Supervisory Senior Counsel (SSC) & Veteran Law Judge (VLJ), I need the ability to designate a case as overtime at the point of assigning to an attorney, so that Attorney productivity/performance can be tracked efficiently. I also need the ability to edit that designation after assignment. 

## Acceptance criteria
- [ ] Please put this work behind the feature toggle: :overtime_revamp
- [ ] This feature should be accessible to the following user groups: VLJs (SSCs are VLJs in Caseflow)
- [ ] Include screenshot(s) in the Github issue if there are front-end changes
- [ ] Update documentation: 
- [ ] When judges or SSCs assign cases to attorneys, they can mark the case as overtime for both AMA and legacy cases
- [ ] Judges or SSCs have the ability to edit the OT designation after assignment
- [ ] Cases marked as OT should display an OT Badge in the Badge Area

### Out of scope
Writing OT designation back to VACOLS from Caseflow

### Current Prototype
Clock icon SVG/HTML can be accessed [here](https://fontawesome.com/icons/clock?style=solid)
[Figma prototype
](https://www.figma.com/proto/qBSLnQhq0H6K7jg6vvW3xW/Mark-AMA-Cases-as-OT?node-id=222%3A0&viewport=254%2C256%2C0.13263243436813354&scaling=min-zoom) ping @geronimoramos if any questions 

### Previous Design mock ups
<details>
<summary>Click to view</summary>
**To mark a case overtime:**
_Senior counsel would click on a blue link in Case Title area of Case Details, reading "Mark this case as overtime":_

![image](https://user-images.githubusercontent.com/59883609/76005663-55494200-5ed9-11ea-8ad5-746ebcc7cf63.png)

_Senior counsel can see a confirmation that this case is now marked overtime. They can also see a new badge appear at the top right of case title.:_
![image](https://user-images.githubusercontent.com/59883609/76005769-7b6ee200-5ed9-11ea-8824-39c43e81aa2d.png)

(The senior counsel will now see another link in Case Title - this time to remove overtime status (in case they made a mistake or have to change the status for other reasons)

_Senior counsel can remove overtime status by clicking on "Remove overtime status for this case"
overtime-marking-remove_
![image](https://user-images.githubusercontent.com/59883609/76005928-b83ad900-5ed9-11ea-8cfc-1978544a4d8c.png)

they'll be prompted with a pop-up modal message asking them "Are you sure you want to remove overtime status?" ----> if they click "yes" they will see a confirmation reading "Vera Marshall's case is no longer marked as overtime. The "OT" status badge has been removed.
-they will also see that the overtime badge has disappeared, and the blue link now reads "Mark this case as overtime"
</details>

### Technical notes
We are currently only tracking when a case was marked as being worked on OT at the time of the attorney. Any changes to OT-related data will need to be captured.

Overtime data is sent to the following Tableau reporting:
The Attorney production and DAS ratings and comments workbooks

Engineering should inquire with the team that will actually do the work to update Tableau reports what format the OT data should be formatted in.

### Resources/other links
Product Brief: https://docs.google.com/document/d/1OayEvH7KAAojD9XVHkfUHyLdyECC-PkeIsrEx9krqDA/edit#

**Comment by hschallhorn** [comment]
Did some [research](https://hackmd.io/@hschallh/SJAYxCC4I) into how we already handle overtime cases in caseflow!

## Investigation

### AMA Cases

#### What is getting sent to the backend?

After selecting "This work product is overtime" and submitting, the following payload is sent to /case_reviews/603/complete

```jsonld
{
  tasks: {
    type: "AttorneyCaseReview",
    work_product: "Decision",
    document_id: "12345-12345678",
    note: "afgdfgfg",
    overtime: true,
    untimely_evidence: false,
    reviewing_judge_id: 3,
    document_type: "draft_decision",
    issues: [{...}]
  }
}
```
Overtime is indeed [something we expect](https://github.com/department-of-veterans-affairs/caseflow/blob/4b7ca54f05d592ef7d635d86463cefa2f12a92d1/app/controllers/case_reviews_controller.rb#L53) to come to the back end. This gets sent to [`CompleteCaseReview`](https://github.com/department-of-veterans-affairs/caseflow/blob/4b7ca54f05d592ef7d635d86463cefa2f12a92d1/app/controllers/case_reviews_controller.rb#L13), which [completes the `AttorneyCaseReview`](https://github.com/department-of-veterans-affairs/caseflow/blob/b2957a98228c406630d30174f475ae2a913ccc1b/app/workflows/complete_case_review.rb#L24). This creates an attorney case review with the provided params (which includes overtime) calls [`update_in_caseflow!`](https://github.com/department-of-veterans-affairs/caseflow/blob/43562ed9e75b14c6f949802521da9df6a4214c2b/app/models/attorney_case_review.rb#L32), which completes the `AttorneyTask`, updates the reviewing judge if changed, and updates issue dispositions. This does not appear to utilize the overtime parameter at all.

#### TLDR
>- [x] What does this actually do for ama cases?
>Sets overtime to "true" for the `AttorneyTask`s associated `AttorneyCaseReview`
>- [x] Can this be leveraged for work on [#13415](https://github.com/department-of-veterans-affairs/caseflow/issues/13415)?
> **NO**. the attorney case review will not be available at judge assign. This needs to be added to the appeal itself

### Legacy Cases
This traces the same steps with similar params

The steps diverge when we call ['update_in_vacols!'](https://github.com/department-of-veterans-affairs/caseflow/blob/43562ed9e75b14c6f949802521da9df6a4214c2b/app/models/attorney_case_review.rb#L23), which reassigns the case to the judge in decass and updates the issue dispositions in vacols. The overtime param IS [used in the reassign](https://github.com/department-of-veterans-affairs/caseflow/blob/43562ed9e75b14c6f949802521da9df6a4214c2b/app/models/attorney_case_review.rb#L57), so let's jump into that. `reassign_case_to_judge` maps the provided params to their [decass values](https://github.com/department-of-veterans-affairs/caseflow/blob/1405be2e9fde453737ffcb286e3a19f516e1a162/app/repositories/queue_repository.rb#L207) and [updates the associated decass record](https://github.com/department-of-veterans-affairs/caseflow/blob/1405be2e9fde453737ffcb286e3a19f516e1a162/app/repositories/queue_repository.rb#L209) with the provided attributes (which include overtime). This uses a combination of the [work product and overtime being true](https://github.com/department-of-veterans-affairs/caseflow/blob/45f40c3aa4b83f90fbebe228cc6052b74702e051/app/mappers/queue_mapper.rb#L143) to determine what value to send to decass as [`deprod`](https://github.com/department-of-veterans-affairs/caseflow/blob/45f40c3aa4b83f90fbebe228cc6052b74702e051/app/mappers/queue_mapper.rb#L6).

#### TLDR
>- [x] Because Judges would also like to mark/unmark legacy cases as overtime (without stipulating which type of overtime) what disposition do we send to vacols now, if any?
>Overtime gets mapped to its correct value based on the work product
>- [x] Can this defaultish value be used for work on [#13415](https://github.com/department-of-veterans-affairs/caseflow/issues/13415)?
>**YES** No default value is used, we can simply pull the `deprod` value and toggle it between its overtime and non overtime version!

**Comment by lomky** [comment]
_[what is this chart?](https://github.com/department-of-veterans-affairs/caseflow/issues/11533)_
```bash
1 | 
2 | |||
3 | |||||||
5 | 
8 | 
```

Removing OT capability for attys in another ticket.
Add OT badge and "Mark/Remove overtime..." to the UI.


Timebox to a 3.

Why 1?

Why 2?
 - Backend already exists, badge area already exists

Why 3?
 - Because FE and new tests
 - Creating a new badge
 
Why 5?

**Comment by geronimoramos** [comment]
Did we want to add this to this ticket? Or create a separate one? 
Add OT badge next to cases that have been assigned overtime status on the case assignment page

![Badge in case assignment page](https://user-images.githubusercontent.com/49766982/77677158-8ce16200-6f4c-11ea-851c-eea70659babf.png)

**Comment by araposo-tistatech** [comment]
@geronimoramos I believe that request is [here](https://github.com/department-of-veterans-affairs/caseflow/issues/13417), but let me know if I missed anything.

**Comment by geronimoramos** [comment]
OT badge color specs:
/* Gray-dark */

background: #323A45;
border-radius: 4px;

**Comment by yoomlam** [comment]
@araposo-tistatech Based on Hunter's investigation comment on Mar 5th, Caseflow "updates the associated decass record with the provided attributes (which include overtime)". This seems to conflict with this ticket description “Out of scope: Writing OT designation back to VACOLS from Caseflow”. Please clarify.

**Comment by araposo-tistatech** [comment]
@yoomlam regrets the Board flagged it would not need to be written back to VACOLS, but I suspect they only meant that for AMA cases and not Legacy. 

Yes, move forward with updating VACOLS as it does today.

**Comment by geronimoramos** [comment]
> @geronimoramos Once OT is set by the VLJ, is the OT badge visible by everyone?

All judges, attorneys, and SSCs from what I recall @araposo-tistatech gathered from stakeholders.

**Comment by geronimoramos** [comment]
**Table** for @geronimoramos 
Context: badge concept breaks the usage/behavior of badges
1. how OT status shows up on queue
2. how OT status shows up on Case Details page. 

In actions drop-down breaks behavior as well because doing tasks from actions dropdown complete or moves that task. 

Thoughts/idea for the future:
1. If users want to modify a characteristic of the case versus a task, something to think about in terms of buttons or dropdowns and placement in Case Details page.

**Comment by yoomlam** [comment]
While we can discuss the [current options for backend changes](https://hackmd.io/doSe3YIFTTiILZ68uHXc9A), the results of user research may reveal new options to be considered or eliminate current options. Kicking this back to design @geronimoramos to clarify the user requirements, which can affect this ticket's AC.

**Comment by yoomlam** [comment]
> @yoomlam regrets the Board flagged it would not need to be written back to VACOLS, but I suspect they only meant that for AMA cases and not Legacy.
> Yes, move forward with updating VACOLS as it does today.

VACOLS continues to be updated with overtime information [here](https://github.com/department-of-veterans-affairs/caseflow/blob/43562ed9e75b14c6f949802521da9df6a4214c2b/app/models/attorney_case_review.rb#L57).
This behavior was not changed by the completion of this ticket.
